simple-request-response/README.md

```diff
@@ -0,0 +1,75 @@
+
+todo: others welcome to join
+Interest Group: [@mxinden @marten-seemann]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@MarcoPolo]: https://github.com/MarcoPolo
+[@mxinden]: https://github.com/mxinden
+
+## Introduction
+
+Many application protocols map well to request/response semantics where a peer
+makes a request to another peer, and receives a response back. For example,
+requesting a file from a peer and getting the file back. In contrast to
+request/response, stream semantics allow two peers to continously send data
+back and forth with no specific ordering. Both styles have their uses. This
+document focuses on _one_ way of implementing request/response semantics and
+how to map it to stream based transports and an HTTP transport.
+
+## Goals of this document
+
+A goal of this document is to define _a_ simple way to write request/response
+style application protocols that can make use of all libp2p transports
+seamlessly (all stream transports and the HTTP transport). This is only one way
+of writing request/response style protocols, and users may choose to not use it.
+
+Another goal is to be backwards compatible with existing application protocols
+that follow these semantics to make upgrading them easier so that they may take
+advantage of the new HTTP transport without sacrificing backwards compatibility.
+
+The final goal is to define something simple such that it's easy to follow,
+implement, and use.
+
+## The abstraction
+
+This abstraction takes in a blob from the requesting peer (aka client) that
+represents the request and delivers the blob to the responding peer (aka
+server). The server responds with a blob that respresents the response. This
+abstraction is agnostic to what the blobs are or how they are encoded. That's up
+to the application protocol to decide.
+
+## How to run on top of a libp2p stream
+
+Each request and response should happen in a single stream. There MUST NOT be
+pipelining. After sending a request, the client SHOULD close its write side
+(signalling EOF to the peer). After handling the response, the client SHOULD
+close the stream. After sending the response the server SHOULD close the stream.
+
+## How to run on top of an HTTP transport
+
+The client's request is placed in the body of an HTTP POST request. The server
+places it's response in the body of the HTTP response. Headers are unused by the
+application protocol (but may be used by the libp2p implementation to provide
+authentication). The HTTP path used is defined by the server's
+`.well-known/libp2p` HTTP resource (see the [HTTP](../http/README.md) spec for
+more details).
+
+
+## Prior Art
+
+This spec is inspired by existing work to use request response protocols on top
+of libp2p streams including, but not limited to:
+- go-libp2p-kad-dht
+- Identify Push
+- AutoNATV2
+
+This is also inspired by rust-libp2p's [Request/Response
+crate](https://docs.rs/libp2p-request-response/0.25.0/libp2p_request_response/).
+
```

Comment:
I'd like an additional section here `## Implementations considerations` that describes that implementations need to be aware of DoS attacks.

Relying on the other side closing the stream is convenient but in the absence of a TLV encoding, we need to make sure to only read a maximum amount of data from the stream while waiting for EOF.

Reply:
What are the DoS attacks you have in mind?

> We need to make sure to only read a maximum amount of data from the stream

Isn't this an application level protocol consideration? If I'm requesting a file, this can be quite big.

Reply:
> What are the DoS attacks you have in mind?

If I rely on the stream being closed before considering the message as received, a malicious peer can send me data until I run out of memory to allocate.

This is how we solve this in `rust-libp2p`: https://github.com/libp2p/rust-libp2p/blob/d3b4714730cafb850be46cbbb7038658d0d560bb/protocols/request-response/src/cbor.rs#L90-L94

> If I'm requesting a file, this can be quite big.

Yes, but there still needs to be an upper limit.

Reply:
I added a note to the doc. This should be defined by the application, so I made a section called "Considerations for applications".

Reply:
Thanks, this is resolved for me (I can't mark it as such).